openage/convert/opus/opusenc.pyx

```diff
@@ -0,0 +1,430 @@
+
+    dbg("Wavefile: Total length: {}\tHeader length: {}\tTrailer length: {}"\
+            .format(len(inputdata), inopt['header_len'], inopt['trailer_len']))
+    if inopt['trailer_len']:
+        inputdata = inputdata[inopt['header_len']:-inopt['trailer_len']]
+    else:
+        inputdata = inputdata[inopt['header_len']:]
+
+    outdata = bytearray(1024*50)
+    cdef size_t outsize = 0  # valid bytes inside 'outdata'
+
+    cdef ogg.ogg_packet op
+    cdef ogg.ogg_page og
+    cdef ogg.ogg_stream_state os
+    cdef opus.OpusEncoder *oe
+    cdef int err
+    cdef int channels = inopt['channels']
+    cdef int wframe_len = inopt['wframe_len']
+    cdef opus.opus_int32 rate = inopt['input_rate']
+    cdef opus.opus_int32 coding_rate = 48000
+    cdef opus.opus_int32 frame_sz = 960
+    cdef opus.opus_int32 lookahead
+
+    if rate > 24000:
+        coding_rate = 48000
+    elif rate > 16000:
+        coding_rate = 24000
+    elif rate > 12000:
+        coding_rate = 16000
+    elif rate > 8000:
+        coding_rate = 12000
+    else:
+        coding_rate = 8000
+
+    frame_sz = <opus.opus_int32> (frame_sz / (48000 / coding_rate))
+
+    if coding_rate != rate:
+        dbg("Resampling necessary: input rate: {}\tcoding rate: {}".format(
+            rate, coding_rate))
+        inputdata = upsample(inputdata, inopt, coding_rate)
+        if not isinstance(inputdata, (bytes, bytearray)):
+            return "Upsampling failed: " + inputdata
+        dbg("Resampled data length: {}".format(len(inputdata)))
+
+    # initialize structs
+    if ogg.ogg_stream_init(&os, int(time.time() * 100) % (1<<31)):
+        return "Could not initialize ogg_stream."
+
+    oe = opus.opus_encoder_create(coding_rate, channels, opus.OPUS_APPLICATION_AUDIO, &err)
+    if err != opus.OPUS_OK:
+        ogg.ogg_stream_clear(&os)
+        if err == opus.OPUS_ALLOC_FAIL:
+            raise MemoryError("Could not allocate opus encoder.")
+        return "Could not allocate opus encoder."
+
+    err = opus.opus_encoder_ctl(oe, opus.OPUS_GET_LOOKAHEAD(&lookahead))
+    if err != opus.OPUS_OK:
+        free_buffers(&os, oe, NULL, NULL)
+        return "Setting encoder settings failed: {}".format(err)
+
+    inopt['pre_skip'] = lookahead / (48000 / coding_rate)
+
+    err = write_opus_header(&os, &op, inopt)
+    if err:
+        free_buffers(&os, oe, NULL, NULL)
+        if err == 1:
+            raise MemoryError("Could not create opus header.")
+        return "Could not write opus header."
+
+    while ogg.ogg_stream_flush(&os, &og):
+        outsize = write_ogg_page(outsize, outdata, &og)
+        if outsize == 0:
+            free_buffers(&os, oe, NULL, NULL)
+            raise MemoryError("Could not write opus header.")
+
+    err = write_opus_comment(&os, &op)
+    if err:
+        free_buffers(&os, oe, NULL, NULL)
+        if err == 1:
+            raise MemoryError("Could not create opus comment.")
+        return "Could not write opus comment."
```

Comment:
I think if you use `ctypedef char[] chararray` and then use `make_unique[chararray]` it could work. But I can't tell why it won't work, sorry.
If the smart pointer is too much effort, I think it's okay to stick with the `bytearray` object from python, it is refcounted and free'd automatically.